scipy/conftest.py

```diff
@@ -39,6 +39,30 @@ def pytest_runtest_setup(item):
     if mark is not None and np.intp(0).itemsize < 8:
         pytest.xfail('Fails on our 32-bit test platform(s): %s' % (mark.args[0],))
 
+    try:
+        from threadpoolctl import threadpool_limits
+        HAS_THREADPOOLCTL = True
+    except (ImportError, AttributeError):
+        # Optional dependency only
+        HAS_THREADPOOLCTL = False
+
+    if HAS_THREADPOOLCTL:
+        # Set the number of openmp threads based on the number of workers
+        # xdist is using to prevent oversubscription. Simplified version of what
+        # sklearn does (it can rely on threadpoolctl and its builtin OpenMP helper
+        # functions)
+        try:
+            xdist_worker_count = int(os.environ['PYTEST_XDIST_WORKER_COUNT'])
+        except KeyError:
+            # raises when pytest-xdist is not installed
+            return
+
+        if not os.getenv('OMP_NUM_THREADS'):
+            max_openmp_threads = os.cpu_count() // 2  # use nr of physical cores
```

Comment:
`os.cpu_count` returns the number of the system CPUs.
In some context, the process might be bound to only some of them.

For instance, you can see it in action using `taskset(1)`:
```sh
$ taskset -c 0-1 python -c "import os; print(os.cpu_count())"
8
```

One can use `os.sched_getaffinity`, which return the list of the usable CPUs.
```sh
$ taskset -c 0-1 python -c "import os; print(len(os.sched_getaffinity(0)))"
2
```

Might this be relevant here?

Reply:
Good question. Sounds like it could be, in particular on HPC systems. I stripped this implementation down from scikit-learn, which uses `joblib.cpu_count` that in the end is: https://github.com/joblib/loky/blob/6245bb2f7c16e283ced3d2ff25a9c9cc439a8867/loky/backend/context.py#L110. From further down in that file, `sched_getaffinity` is checked for but not always available.

This is a pretty deep rabbit hole, so I'm inclined to leave this as is - solves the problem for 95%. If we want to solve it for ~100%, then I'd prefer to add an optional dependency on `joblib` / `loky` (or vendor it), rather than try and reimplement it.

Reply:
For reference the actual implementation is there:

https://github.com/joblib/loky/blob/6245bb2f7c16e283ced3d2ff25a9c9cc439a8867/loky/backend/context.py#L110-L255

It takes both CPU affinity and CFS CPU time quotas into account to compute the rounded actual amount of CPU allowed for the user's processes. It also has an option to count only physical cores instead of logical cores (hyperthreads) when no such restrictions are in place. It is quite often the case that hyperthreads are pretty useless for OpenBLAS.

Reply:
Thanks @ogrisel!

> It is quite often the case that hyperthreads are pretty useless for OpenBLAS.

Yes, I've noticed that too. The performance is only marginally better typically, and sometimes it can even degrade.